knodle/trainer/crossweigh_weighing/data_splitting_by_rules.py

```diff
@@ -0,0 +1,289 @@
+
+    :param data_features: numpy array with encoded data samples (num_samples, num_features)
+    :param rules_ids: list of shuffled rules indices
+    :param rule_id2samples_ids: dictionary that contains information about corresponding between rules ids and sample ids.
+    Rule in this context means anything basing on which splitting is performed (matched rule, sample signature, ...)
+    :param labels: labels of all training samples (num_samples, num_classes)
+    :param fold: number of a current hold-out fold
+    :param num_folds: the whole number of folds the data should be splitted into
+    :param other_sample_ids: a list of sample ids that belong to the other class. They won't be included in the test
+    set, but only to the training set.
+    should be
+
+    :return: dataloaders for cw training and testing
+    """
+    train_rules_idx, test_rules_idx = calculate_rules_indices(rules_ids, fold, num_folds)
+
+    # select train and test samples and labels according to the selected rules idx
+    test_samples, test_labels, test_idx = get_samples_labels_idx_by_rule_id(
+        data_features, labels, test_rules_idx, rule_id2samples_ids, check_intersections=None,
+        other_sample_ids=other_sample_ids
+    )
+
+    train_samples, train_labels, _ = get_samples_labels_idx_by_rule_id(
+        data_features, labels, train_rules_idx, rule_id2samples_ids, check_intersections=test_idx,
+        other_sample_ids=other_sample_ids
+    )
+
+    train_dataset = input_labels_to_tensordataset(train_samples, train_labels)
+    test_dataset = input_info_labels_to_tensordataset(test_samples, test_idx, test_labels)
+
+    logger.info(
+        f"Fold {fold}     Rules in training set: {len(train_rules_idx)}, rules in test set: {len(test_rules_idx)}, "
+        f"samples in training set: {len(train_samples)}, samples in test set: {len(test_samples)}"
+    )
+
+    return train_dataset, test_dataset
+
+
+def calculate_rules_indices(rules_idx: list, fold: int, num_folds: int) -> (np.ndarray, np.ndarray):
+    """
+    Calculates the indices of the samples which are to be included in training and test sets for k-fold cross validation.
+
+    :param rules_idx: all rules indices (shuffled) that are to be splitted into cw training & cw test set rules
+    :param fold: number of a current hold-out fold
+
+    :return: two arrays containing indices of rules that will be used for cw training and cw test set accordingly
+    """
+    test_rules_idx = rules_idx[fold::num_folds]
+    train_rules_idx = [rule_id for rule_id in rules_idx if rule_id not in test_rules_idx]
+
+    if not set(test_rules_idx).isdisjoint(set(train_rules_idx)):
+        raise ValueError("Splitting into train and test rules is done incorrectly.")
+
+    return train_rules_idx, test_rules_idx
+
+
+def get_samples_labels_idx_by_rule_id(
+        data_features: np.ndarray, labels: np.ndarray, indices: list, rules_samples_ids_dict: Dict,
```

Comment:
also rule2sample_ids? (these suggestion are not constistent, I probably missed a couple of spots :D)